docs/proposals/advanced-nginx-extensions.md

```diff
@@ -0,0 +1,28 @@
+# Enhancement Proposal-2035: Advanced NGINX Extensions
+
+- Issue: https://github.com/nginxinc/nginx-gateway-fabric/issues/2035
+- Status: Provisional
+
+## Summary
+
+NGINX Gateway Fabric (NGF) [exposes](/site/content/overview/gateway-api-compatibility.md) NGINX features via Gateway API
+resources (like HTTPRoute) and [NGINX extensions](nginx-extensions.md) (like ClientSettingPolicy). Combined, they
+expose a subset of the most common NGINX configuration. As we implement more Gateway API resources and NGINX extensions,
+the subset will grow. However, it will take time. Additionally, because the number of NGINX configuration directives
+and parameters is huge, not all of them will be supported that way. As a result, users are not able to implement certain
+NGINX use cases. To allow them to implement those use cases, we need to bring a new extension mechanism to NGF.
+
+## Goals
+
+- Allow users to insert NGINX configuration not supported via Gateway API resources or NGINX extensions.
+- Allow users to customize supported NGINX configuration (for example, add a parameter to an NGINX directive).
+- Support configuration from modules not loaded in NGINX by default or third-party modules.
+- Most of the configuration complexity should fall onto the cluster operator persona, not the application developer.
+- Provide security controls to prevent application developers from injecting arbitrary NGINX configuration.
+- Ensure adequate configuration validation to prevent NGINX outages due to invalid configuration.
+- Don't require building a custom NGF binary (source code modifications).
+
+## Non-Goals
+
+- Support configuration other than NGINX directives. For example, njs configuration files or TLS certificates.
+- Reimplement already supported features through the new extension mechanism.
```

Comment:
This will almost certainly happen, correct? Every first-class feature we make in our resources can also likely be replicated using this advanced NGINX extensions feature, right?

Reply:
> Every first-class feature we make in our resources can also likely be replicated using this advanced NGINX extensions feature, right?

More or less.

However, some features will require extra configuration resources - ex. specifying auth keys in secrets. For example, like ConfigMap in backendTLSPolicy -- https://docs.nginx.com/nginx-gateway-fabric/how-to/traffic-management/securing-backend-traffic/#create-the-backend-tls-configuration 

Supporting extra resources like ConfigMap in backendTLSPolicy is out of scope.

> This will almost certainly happen, correct? 

This  proposal doesn't plan to re-implement existing features using the new extension mechanism. For example, we just added ClientSettingPolicy. For that, we wrote some go code in https://github.com/nginxinc/nginx-gateway-fabric/pull/2013/files Potentially, the new extension mechanism will allow developing a similar policy and adding it to NGF without writing any go code, so that ClientSettingPolicy can be reimplemented and the go code can be removed. However, this proposal doesn't pursue that goal.

At the same time, users can do whatever they want :)